### PR TITLE
[mono] Precompute the CallInfo structure used by the mono_arch_..nati…

### PR DESCRIPTION
**Pull request by vargaz**
…ve_call_context_.. functions.

### CODE REVIEW

**Comment by vargaz** [comment]
This speeds up ves_pinvoke_method () in the interpreter, including the icalls which use QCallTypeHandle.

File: src/mono/mono/mini/interp/interp.c [hunk 1]
```diff
@@ -1680,8 +1680,15 @@ ves_pinvoke_method (
 	}
 
 #ifdef MONO_ARCH_HAVE_INTERP_PINVOKE_TRAMP
+	gpointer call_info = *cache;
+
+	if (!call_info) {
+		call_info = mono_arch_get_interp_native_call_info (get_default_mem_manager (), sig);
+		mono_memory_barrier ();
+		*cache = call_info;
```
Comment (BrzVlad) [inline src/mono/mono/mini/interp/interp.c:1688]: We should use a CAS here to avoid leaks from races

File: src/mono/mono/mini/mini-amd64.c [hunk 2]
```diff
@@ -1178,11 +1187,33 @@ arg_set_val (CallContext *ccontext, ArgInfo *ainfo, gpointer src)
 	}
 }
 
+gpointer
+mono_arch_get_interp_native_call_info (MonoMemoryManager *mem_manager, MonoMethodSignature *sig)
+{
+    CallInfo *cinfo = get_call_info (NULL, sig);
```
Comment (BrzVlad) [inline src/mono/mono/mini/mini-amd64.c:1193]: whitespace damage

File: src/mono/mono/mini/mini-arm.c [hunk 3]
```diff
@@ -1666,12 +1675,34 @@ arg_set_val (CallContext *ccontext, ArgInfo *ainfo, gpointer src)
 	memcpy (ccontext->stack + ainfo->offset, (host_mgreg_t*)src + ainfo->size, ainfo->struct_size - reg_size);
 }
 
+gpointer
+mono_arch_get_interp_native_call_info (MonoMemoryManager *mem_manager, MonoMethodSignature *sig)
+{
+    CallInfo *cinfo = get_call_info (NULL, sig);
```
Comment (BrzVlad) [inline src/mono/mono/mini/mini-arm.c:1681]: spaces

File: src/mono/mono/mini/mini-arm64.c [hunk 4]
```diff
@@ -1825,12 +1834,34 @@ arg_set_val (CallContext *ccontext, ArgInfo *ainfo, gpointer src)
 	}
 }
 
+gpointer
+mono_arch_get_interp_native_call_info (MonoMemoryManager *mem_manager, MonoMethodSignature *sig)
+{
+    CallInfo *cinfo = get_call_info (NULL, sig);
```
Comment (BrzVlad) [inline src/mono/mono/mini/mini-arm64.c:1840]: spaces

File: src/mono/mono/mini/interp/interp.c (hunk 1 above)
Comment (vargaz) [reply src/mono/mono/mini/interp/interp.c:1688]: The memory is allocated from the mem manager, so its going to be leaked anyway.

Comment (BrzVlad) [reply src/mono/mono/mini/interp/interp.c:1688]: I see. Also a `mono_memory_write_barrier` here is enough.

**Comment by vargaz** [comment]
This speeds up this testcase by about 50% when running with the interpreter:
```
        for (int i = 0; i < 10000000; ++i) {
            var t = typeof (int).DeclaringType;
        }
```

**Review by lewing — APPROVED** [review]

**Pull request merged**